### PR TITLE
refactor: Combine ExtensionSolutions (no separate closure)

### PR DESCRIPTION
**Pull request by acl-cqc**
* `infer::infer_extensions` returns only a combined solution (for previously-open locations), after variables instantiated
* `Hugr::infer_extensions` writes (all parts of) the solution into place *and* returns it
* `validate_with_extension_closure` left in-place, with test demonstrating usage w/ sub-DFGs
* This should open the way (in future PRs) to changing implementation of `infer::infer_extensions`

### CODE REVIEW

**Comment by acl-cqc** [comment]
IIUC coverage is not reduced by the PR, just that the lines affected by the PR have lower-than-average coverage anyway, and AFAICS that is because of `assert_matches!`, so I think this is OK.

File: quantinuum-hugr/src/extension/infer/test.rs [hunk 1]
```diff
@@ -795,6 +794,28 @@ fn test_cfg_loops() -> Result<(), Box<dyn Error>> {
...
+        ExtensionSet::singleton(&B),
+    )?;
+    assert_matches!(
+        hugr.validate(&PRELUDE_REGISTRY),
+        Err(ValidationError::ExtensionError(_))
+    );
+
+    let immut = hugr.clone();
+    let soln = infer_extensions(&immut)?;
+    immut.validate_with_extension_closure(soln, &PRELUDE_REGISTRY)?;
```
Comment (croyzor) [inline quantinuum-hugr/src/extension/infer/test.rs:812]: Looks like this should just be `immut.validate(soln, &PRELUDE_REGISTRY)`?

Comment (acl-cqc) [reply quantinuum-hugr/src/extension/infer/test.rs:813]: ?? `validate` takes only one argument, the registry.

**Comment by croyzor** [comment]
Casting my mind back, I think the motivation for returning the open solution is that we might want to run extension inference on a subgraph before we add it to the main graph. I'm not really sure when this would be the case -- I think the case is that when doing replacements, we run inference on the complete graph before and after, so it's probably unnecessary

**Comment by acl-cqc** [comment]
> Casting my mind back, I think the motivation for returning the open solution is that we might want to run extension inference on a subgraph before we add it to the main graph. I'm not really sure when this would be the case -- I think the case is that when doing replacements, we run inference on the complete graph before and after, so it's probably unnecessary

Ah-ha, I see. So the flow would have been:
1. Run inference on the subgraph. Write into the subgraph only the solutions that do not depend on the subgraph inputs.
2. Substitute the subgraph into the main graph.
3. Rerun inference on the main graph in order to figure out the solutions that did depend on the subgraph inputs.

Noting that if in step 1 we wrote in all the solutions, including those dependent on the subgraph inputs, then step 3 would fail because the inserted graph would have bad solutions (e.g. emptyset) that conflict.

This PR removes the ability to write *only* those solutions into the subgraph. However, the flow is largely intact if one does not write *any* solutions into the subgraph before insertion. (You can still validate the subgraph using the full set of solutions, if required.)

Close enough? I think I can realize that in a test.

**Comment by acl-cqc** [comment]
@croyzor added a much more thorough test which I hope covers that use-case (as well as we can - there'll be a better, but separate, solution if we do #709). I had to make `infer_extensions(&Hugr)` return only the *new* solutions rather than existing ones to get around assertions in `gather_extensions`. Can you take another look?

**Review by croyzor — APPROVED** [review]
Nice, let's update this once can give a better answer for polymorphic hugrs!

**Pull request merged**